src/expr/src/relation/func.rs

```diff
@@ -616,27 +645,30 @@ fn lag_lead_inner_ignore_nulls<'a>(
                 // We start j from idx, and step j until we have seen an abs(offset) number of non-null
                 // values or reach the beginning or end of the partition.
                 //
-                // If offset is big, then this is slow: `O(partition_size * offset)`. We could avoid an
-                // inner loop, and instead step two indexes in one loop, with one index lagging behind.
-                // But a common use case is an offset of 1, for which this doesn't matter.
+                // If offset is big, then this is slow: Considering the entire function, it's
+                // `O(partition_size * offset)`.
+                // However, a common use case is an offset of 1, for which this doesn't matter.
+                // TODO: For larger offsets, we could have a completely different implementation
+                // that steps two indexes in one loop, with one index lagging behind.
```

Comment:
Can you explain the two-indexes approach, I'm not smart enough to see how it works!

Also, why did you decide to go with the skiplist approach rather than implementing this presumably better one? Is it more complicated or less efficient for smaller offsets?

Reply:
The trick would be to remember where we found the previous result (let's called it `prev_lagged_idx`), and start stepping the inner loop not from `idx` but from `prev_lagged_idx`. The number of steps that we need to do in the inner loop would be
`idx - prev_idx + (offset - prev_offset) * lag_lead_sign`. (See a WIP branch [here](https://github.com/ggevay/materialize/commit/0baab443d07f673e13ea13c29daf253dff26736f).)
If the offsets don't jump around wildly, then this works well for both small and big offsets. Specifically, for a constant offset, the run time of the whole function would be `O(partition_size)` rather than the current `O(partition_size * offset)`.

At first glance, the above sounds not much more complicated than the current implementation, but unfortunately there is a tricky complication. I've implemented the above on the above branch, but it has a big bug. At the moment, I don't see how to fix it without a similar precomputation as the `skip_nulls_forward` and `skip_nulls_backward` of the current approach. The problem is that when `to_go` is stepping in the direction inwards to `idx`, then we have to specially handle the case when we reach `idx` and see a null there. This is because the `idx` position is special: a normal inner loop (such as what the current implementation does) would _start_ stepping from there outwards, so if we skip over it in the inwards direction then we end up in a state where a normal inner loop wouldn't be, because it would correspond to a negative number of steps of the normal inner loop.

For example, let's say that we are computing `lag(1) IGNORE NULLS`, and the input values are
```
a, b, null, null, c
0  1  2     3     4
```
where the second line simply shows the indexes, and we are at the point where we want to compute a result for the second null, so
```
idx = 3
prev_idx = 2
offset = 1
prev_offset = 1
prev_lagged_idx = 1
```
So, we start `j` from 1, and step up towards 2. But we find a null there, so we step over the nulls and reach 4, and wrongly give `c` as the result, when in fact the result would still be `b` (like in the previous step of the outer loop).

To solve this problem, my only idea at the moment would need `skip_nulls_forward`, because otherwise we detect the problematic case after doing `O(null_run_length)` steps of the inner loop, and we'd do this at every step of the outer loop while we are in the same run of nulls, so overall we'd again be `O(null_run_length^2)`.

I'd say we can hold off on implementing this (or figuring out a better way) until somebody shows up with a performance problem caused by a large offset.